3802 reconsent script

pfriedman opened the pull request:
The basic rake task to create a Reconsent Informed Consent Event in Cases after having scheduled the stand-alone event in PSC.

Commit:
Moved date available check to participant #3802

so that the check can be used by the reconsent scheduling rake task

Commit:
First pass at rake task #3802

to schedule re-consents

Commit:
specs for date_available #3802

rsutphin commented on app/models/participant.rb:
```diff
@@ -954,6 +954,24 @@ def should_be_screened?
+      !ics_dates.include?(dt)
+    rescue ArgumentError
+      # if date cannot be parsed do not allow user to schedule the informed consent event
+      false
```
Input parsing should be handled at the layer which can report the error to the user. This method should only accept a `Date`.

rsutphin commented on app/models/participant.rb:
```diff
@@ -954,6 +954,24 @@ def should_be_screened?
+  # @param date [String]
+  # @return [Boolean]
+  def date_available_for_informed_consent_event?(date)
+    return true if events.blank?
```
This is unnecessary. If there are no events, the main body of the method will return true.

To put it another way, you might have a guard like this to allow you to make a simplifying assumption in later code. In this case, there is no such simplification granted; this guard is just increased complexity.

rsutphin commented on lib/tasks/import.rake:
```diff
@@ -280,28 +280,40 @@ namespace :import do
 
+  desc 'Schedule Re-consent Event for a particular date for all followed Participants'
+  task :schedule_reconsent => [:psc_setup, :environment, :set_whodunnit, :find_followed_participants_for_psc] do
+    date = Date.new(2013, 1, 1)
```
Do we really want to hard-code this date? My understanding was that it would be different for different locations.

rsutphin commented on lib/tasks/import.rake:
```diff
@@ -280,28 +280,40 @@ namespace :import do
   task :cancel_activities_with_non_matching_mdes_instruments => [:psc_setup, :environment, :set_whodunnit]  do
     non_children_participants.each do |part|
-      msg = "Looking for activities to cancel for participant #{part.p_id}..."
+      msg = "Looking for activities to cancel for participant #{part.p_id}...\n"
```
This is the wrong place to fix this. You want a newline in the terminal output, yes, but not in the rails log. `msg` gets sent to both. Use `$stderr.puts(msg)` (instead of `$stderr.print`) to get a newline in the terminal without adding blank lines to the rails log.

The only reason why you should use `$stderr.print` is if you are doing something like:

``` ruby
$stderr.print "Doing a slow thing..."
# do slow thing
$stderr.puts "done."
```

This will print both messages on the same line, which is nice. But generally, for terminal output `puts` is what you want unless you are deliberately being fancy.

rsutphin commented on lib/tasks/import.rake:
```diff
@@ -280,28 +280,40 @@ namespace :import do
+  desc 'Schedule Re-consent Event for a particular date for all followed Participants'
+  task :schedule_reconsent => [:psc_setup, :environment, :set_whodunnit, :find_followed_participants_for_psc] do
+    date = Date.new(2013, 1, 1)
+    expected_followed_participants_for_psc.each do |pt|
```
Are we reconsenting mothers only, or children too? `expected_followed_participants_for_psc` will only contain mothers. Or does the reconsent workflow allow for both?

pfriedman replied:
This task just schedules the re-consent for the mother as they own the schedule in PSC and was easiest to test if this works. I'll follow up with those in contact with the data collectors if we should also schedule the re-consent activities for the children.

rsutphin commented on lib/tasks/import.rake:
```diff
@@ -280,28 +280,40 @@ namespace :import do
+        msg = resp.success? ? "Re-consent event scheduled for #{pt.p_id}.\n" : "!!! Could not schedule informed consent for #{p.p_id} !!!\n"
+        $stderr.print(msg)
+        Rails.logger.info(msg)
+      end
```
This task is going to silently do nothing if the participant already has an informed consent on the selected date. Is that the right thing to do?

Commit:
Update based on feedback #3802

The Participant#date_available_for_informed_consent_event?
method should only take a Date (not either a Date or String) so
update this method and callers accordingly.

This date can be given by the user calling the rake task to
schedule reconsent events via highline.

In all places attempting to schedule a participant notify the
user of the progress or of any issues.
